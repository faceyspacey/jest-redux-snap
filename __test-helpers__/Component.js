import React from 'react'
import { connect } from 'react-redux'

const Component = ({  num, empty }) =>
  <div>
    <span>{num}</span>
    {!empty && <span>something</span>}
  </div>

const mapState = ({ num, empty }) => ({ num, empty })

export default connect(mapState)(Component)
