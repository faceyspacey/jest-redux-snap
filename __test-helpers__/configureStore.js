import { createStore } from 'redux'


export default () => {
  const reducer = (state = { num: 0, empty: false }, action = {}) => {
    switch (action.type) {
      case 'INCREMENT':
        return {
          ...state,
          num: state.num + 1,
        }
      case 'EMPTY':
        return {
          ...state,
          empty: true,
        }
      case 'FULL':
        return {
          ...state,
          empty: false,
        }
      default:
        return state
    }
  }

  return createStore(reducer)
}
