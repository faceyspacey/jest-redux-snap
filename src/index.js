export { default as snap } from './snap'
export { default as createApp } from './createApp'
