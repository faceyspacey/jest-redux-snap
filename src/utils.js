// eg: MyComponent
export const isClass = Component => typeof Component === 'function' && !isElement(Component) && !isInstance(Component)

// eg: renderer.create(<MyComponent />).toJSON()
export const isTree = Component => Component && Component.$$typeof && Component.$$typeof.toString() === 'Symbol(react.test.json)'

// eg: <MyComponent />
export const isElement = Component => Component && Component.$$typeof

// eg: renderer.create(<MyComponent />)
export const isInstance = Component => Component && Component.toJSON


export const isStore = arg => typeof arg === 'object' && arg.getState
