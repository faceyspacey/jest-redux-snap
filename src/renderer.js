import { isInstance } from './utils'


export function DeepRenderer(instance, tree) {
  this._instance = instance
  this._tree = tree
}

DeepRenderer.prototype = {
  // calling tree() insures having up to date instance in response to state changes
  tree() {
    return isInstance(this._instance) ? this._instance.toJSON() : this._tree
  },

  // calling snap will of course also use most up to date instance
  snap() {
    expect(this.tree()).toMatchSnapshot()
    return this
  },

  // this._instance can be passed to snap(renderer.instance()) -- this is as opposed to doing this.snap()
  component() {
    return this._instance
  },
}


export function ShallowRenderer(renderer, element, tree) {
  this._renderer = renderer
  this._element = element
  this._tree = tree
}

ShallowRenderer.prototype = {
  tree() {
    return this._element ? this._renderer.getRenderOutput() : this._tree
  },
  snap() {
    expect(this.tree()).toMatchSnapshot()
    return this
  },
  component() {
    return this._element
  },
}
