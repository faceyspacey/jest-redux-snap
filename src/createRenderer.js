import React from 'react'
import renderer from 'react-test-renderer'
import ReactTestUtils from 'react-addons-test-utils'

import { isClass, isTree, isElement, isInstance } from './utils'
import { DeepRenderer, ShallowRenderer } from './renderer'


export default function createRenderer(target, props, deep = true) {
  if (typeof props === 'boolean') {
    deep = props
    props = undefined
  }

  if (!deep) {
    return _createShallowRenderer(target, props)
  }

  let tree
  let instance

  if (isClass(target)) {
    instance = renderer.create(<target {...props} />)
  }
  else if (isTree(target)) {
    tree = target
  }
  else if (isElement(target)) {
    instance = renderer.create(target)
  }
  else if (isInstance(target)) {
    instance = target
  }
  else {
    tree = target
  }

  return new DeepRenderer(instance, tree)
}


export function _createShallowRenderer(target, props) {
  const renderer = ReactTestUtils.createRenderer()

  let tree
  let element

  if (isClass(target)) {
    element = <target {...props} />
    renderer.render(element)
  }
  else if (isTree(target)) {
    tree = target
  }
  else if (isElement(target)) {
    element = target
    renderer.render(element)
  }
  else {
    tree = target
  }

  return new ShallowRenderer(renderer, element, tree)
}


// ReactTestUtils does not create component instances that play nice with Jest mocks.
// A mocked component passed to ReactTestUtils.createRenderer().render(component)
// will expect there to be no props passed, which defeats its purpose.

const originalError = global.console.error.bind(global.console)

global.console.error = (...args) => {
  const msg = args[0]
  if (msg && msg.indexOf('Unknown prop') > -1) {
    return
  }

  return originalError(...args)
}
