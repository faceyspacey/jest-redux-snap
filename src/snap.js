import createRenderer from './createRenderer'

export default function snap(target, props, deep = true) {
  return createRenderer(target, props, deep).snap()
}
