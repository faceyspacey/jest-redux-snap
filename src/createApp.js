import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import createRenderer from './createRenderer'
import snap from './snap'


export default (store, ReactComponentClass, mapStateToProps) => {
  let wrapper

  store.subscribe(() => {
    wrapper.forceUpdate()
  })

  const element =
    <Provider store={store}>
      <ReactiveWrapper
        ref={wr => wrapper = wr}
        store={store}
        Component={ReactComponentClass}
        mapStateToProps={mapStateToProps}
      />
    </Provider>

  const instance = renderer.create(element)
  const app = createRenderer(instance)

  addMethods(app, store, element)

  return app
}


class ReactiveWrapper extends React.Component {
  render() {
    const { store, Component, mapStateToProps } = this.props

    const props = typeof mapStateToProps === 'function'
      ? mapStateToProps(store.getState(), store.dispatch)
      : mapStateToProps // object or undefined

    return <Component {...props} />
  }
}


const addMethods = (app, store, element) => {
  app.element = () => element
  app.story = () => element

  app.getState = () => store.getState()
  app.dispatch = action => store.dispatch(action)

  app.snapAction = (action) => {
    if (typeof action !== 'function') {
      snap(action)
    }

    store.dispatch(action)
    snap(store.getState())
    app.snap()
  }

  app.snapState = () => snap(app.getState())
}
