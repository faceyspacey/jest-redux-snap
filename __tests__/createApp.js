import React from 'react'
import renderer from 'react-test-renderer'

import configureStore from '../__test-helpers__/configureStore'
import Component from '../__test-helpers__/ContainerComponent'
import { createApp } from '../src'


describe('createApp(store, Class) ', () => {
  it('app.tree() and app.component().toJSON() updates in response to state changes', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    console.log(app.tree().children)
    app.snap()

    store.dispatch({ type: 'EMPTY' })
    console.log(app.tree().children)
    app.snap()

    store.dispatch({ type: 'INCREMENT' })
    console.log(app.component().toJSON().children)
    app.snap()
  })

  it('app.snap()', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    app.snap()
  })

  it('app.tree()', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    console.log(app.tree())

    expect(app.tree()).toMatchSnapshot()
  })

  it('app.component()', () => {
    const store = configureStore()
    const app = createApp(store, Component)
    const tree = app.component().toJSON()

    expect(tree).toEqual(app.tree())
    expect(tree).toMatchSnapshot()
  })

  it('app.element() [alias: app.story()]', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    const element = renderer.create(app.element()).toJSON()
    const story = renderer.create(app.story()).toJSON()

    console.log(element)
    console.log(story)

    expect(element).toEqual(story)
    expect(app.tree()).toEqual(element)

  })

  it('app.dispatch()', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    app.dispatch({ type: 'EMPTY' })
    app.snap()
  })

  it('app.getState()', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    console.log(app.getState())
    expect(app.getState()).toMatchSnapshot()

    app.dispatch({ type: 'INCREMENT' })

    console.log(app.getState())
    expect(app.getState()).toMatchSnapshot()
  })

  it('app.snapState()', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    app.dispatch({ type: 'INCREMENT' })

    app.snapState()
  })

  it('app.snapAction()', () => {
    const store = configureStore()
    const app = createApp(store, Component)

    app.snapAction({ type: 'INCREMENT' })
  })
})


describe('createApp(store, Class, mapStateToProps)', () => {
  it('mapStateToProps == object', () => {
    const store = configureStore()
    const app = createApp(store, Component, { title: 'FOO' })

    console.log(app.tree().children)
    expect(app.tree().children[0].children[0]).toEqual('FOO')
    expect(app.tree().children[1].children[0]).toEqual(0)
    app.snap()

    app.dispatch({ type: 'INCREMENT' })

    console.log(app.tree().children)
    expect(app.tree().children[0].children[0]).toEqual('FOO')
    expect(app.tree().children[1].children[0]).toEqual(1)
    app.snap()
  })

  it('mapStateToProps == function', () => {
    const store = configureStore()
    const mapState = state => ({ title: `title: ${state.num}` })
    const app = createApp(store, Component, mapState)

    console.log(app.tree().children[0])
    expect(app.tree().children[0].children[0]).toEqual('title: 0')
    expect(app.tree().children[1].children[0]).toEqual(0)
    app.snap()

    app.dispatch({ type: 'INCREMENT' })

    console.log(app.tree().children[0])
    expect(app.tree().children[0].children[0]).toEqual('title: 1')
    expect(app.tree().children[1].children[0]).toEqual(1)
    app.snap()
  })
})
