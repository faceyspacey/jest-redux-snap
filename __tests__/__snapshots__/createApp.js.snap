// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createApp(store, Class)  app.component() 1`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    0
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class)  app.dispatch() 1`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    0
  </span>
</div>
`;

exports[`createApp(store, Class)  app.getState() 1`] = `
Object {
  "empty": false,
  "num": 0,
}
`;

exports[`createApp(store, Class)  app.getState() 2`] = `
Object {
  "empty": false,
  "num": 1,
}
`;

exports[`createApp(store, Class)  app.snap() 1`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    0
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class)  app.snap(action) 1`] = `
Object {
  "type": "INCREMENT",
}
`;

exports[`createApp(store, Class)  app.snap(action) 2`] = `
Object {
  "empty": false,
  "num": 1,
}
`;

exports[`createApp(store, Class)  app.snap(action) 3`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    1
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class)  app.snap(thunkAction) 1`] = `"thunk"`;

exports[`createApp(store, Class)  app.snap(thunkAction) 2`] = `
Object {
  "empty": false,
  "num": 1,
}
`;

exports[`createApp(store, Class)  app.snap(thunkAction) 3`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    1
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class)  app.snapState() 1`] = `
Object {
  "empty": false,
  "num": 1,
}
`;

exports[`createApp(store, Class)  app.tree() 1`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    0
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class)  app.tree() and app.component().toJSON() updates in response to state changes 1`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    0
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class)  app.tree() and app.component().toJSON() updates in response to state changes 2`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    0
  </span>
</div>
`;

exports[`createApp(store, Class)  app.tree() and app.component().toJSON() updates in response to state changes 3`] = `
<div>
  <h1>
    a title
  </h1>
  <span>
    1
  </span>
</div>
`;

exports[`createApp(store, Class, mapStateToProps) mapStateToProps == function 1`] = `
<div>
  <h1>
    title: 0
  </h1>
  <span>
    0
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class, mapStateToProps) mapStateToProps == function 2`] = `
<div>
  <h1>
    title: 1
  </h1>
  <span>
    1
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class, mapStateToProps) mapStateToProps == object 1`] = `
<div>
  <h1>
    FOO
  </h1>
  <span>
    0
  </span>
  <span>
    something
  </span>
</div>
`;

exports[`createApp(store, Class, mapStateToProps) mapStateToProps == object 2`] = `
<div>
  <h1>
    FOO
  </h1>
  <span>
    1
  </span>
  <span>
    something
  </span>
</div>
`;
